# 数据处理

In [268]:
import json


with open('LabeledDataCCG.json', 'r') as f:
    data = json.load(f)

In [269]:
data[0]

{'CIK': '1000228',
 'Charter_ID': '1000228A20180531',
 'Date_Coded': '1997-11-12',
 'Date_Filing': '2007-02-28',
 'CompanyName': 'HENRY SCHEIN INC',
 'State': 'DE',
 'FullRestate': 'N',
 'EntityType': '',
 'MultComm': '',
 'MultCommUnequal': '',
 'MultPreferred': '',
 'PreferredVote': '',
 'PreferredVoteLimited': '',
 'BlankCheck1': '',
 'BustBankSHVote': '',
 'BustBankBoD': '',
 'ReallocateComPref': '',
 'UnequalVote': '',
 'WCterm': '',
 'WCCond': '',
 'WCProhibit': '',
 'SMTerm': '',
 'SMCond': '',
 'SMProhibit': '',
 'StaggeredBd': '',
 'StaggeredBdPhaseOut': '',
 'CumulativeVoting': '',
 'CharterSupMaj': 'Partly',
 'MergerSupMaj': '',
 'CorpOppDirector': '',
 'CorpOppOfficer': '',
 'CorpOppSH': '',
 'Directors102b7': '',
 'Officers102b7': '',
 'IndemnityDirector': '',
 'IndemnityOfficer': '',
 'AdvancementDirector': '',
 'AdvancementOfficer': '',
 'InsuranceDirector': '',
 'InsuranceOfficer': ''}

In [270]:

fields_full = [
    "CIK",
    "Charter_ID",
    "Date_Coded",
    "Date_Filing",
    "CompanyName",
    "State",
    "FullRestate",
    "EntityType",
    "MultComm",
    "MultCommUnequal",
    "MultPreferred",
    "PreferredVote",
    "PreferredVoteLimited",
    "BlankCheck1",
    "BustBankSHVote",
    "BustBankBoD",
    "ReallocateComPref",
    "UnequalVote",
    "WCterm",
    "WCCond",
    "WCProhibit",
    "SMTerm",
    "SMCond",
    "SMProhibit",
    "StaggeredBd",
    "StaggeredBdPhaseOut",
    "CumulativeVoting",
    "CharterSupMaj",
    "MergerSupMaj",
    "CorpOppDirector",
    "CorpOppOfficer",
    "CorpOppSH",
    "Directors102b7",
    "Officers102b7",
    "IndemnityDirector",
    "IndemnityOfficer",
    "AdvancementDirector",
    "AdvancementOfficer",
    "InsuranceDirector",
    "InsuranceOfficer",
]
fields = [
    "CIK",
    "Charter_ID",
    "Date_Coded",
    "Date_Filing",
    "CompanyName",
    "State",
    # Governance-related fields of interest
    #"StaggeredBd",
    "CumulativeVoting",
    #"CharterSupMaj",
    #"CorpOppDirector",
    "Directors102b7",
    "Officers102b7",
    "IndemnityDirector",
    "IndemnityOfficer",
    # Optionally include other fields if they are relevant to your analysis
    #"AdvancementDirector",
    # "AdvancementOfficer",
    #"InsuranceDirector",
    # "InsuranceOfficer",
]

#data = [{k:d[k] for k in fields} for d in data]
#data = [{k: d[k] for k in fields if d[k] != 'NAN'} for d in data if all(d[field] != 'NAN' for field in fields)]
data = [
    {k: d[k] for k in fields if d[k] not in ['NAN', '']} 
    for d in data 
    if all(d[field] not in ['NAN', ''] for field in fields)
]

In [271]:
import random
from collections import Counter

frac = 0.05
random.shuffle(data)

data_sub = data[:int(len(data)*frac)]
print(f'Amount:{len(data_sub)}')
Counter([i['Directors102b7'] for i in data_sub])

Amount:246


Counter({'Y': 205, 'N': 41})

# code

In [272]:
import os

os.environ['OPENAI_API_KEY'] = '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'

In [273]:
from langchain_openai import ChatOpenAI
from langchain_core.prompts import ChatPromptTemplate, PromptTemplate

In [274]:
from openai import Client

model_list = Client().models.list()
model_list.data

[Model(id='dall-e-3', created=1698785189, object='model', owned_by='system'),
 Model(id='dall-e-2', created=1698798177, object='model', owned_by='system'),
 Model(id='gpt-3.5-turbo-0125', created=1706048358, object='model', owned_by='system'),
 Model(id='text-embedding-ada-002', created=1671217299, object='model', owned_by='openai-internal'),
 Model(id='tts-1-hd-1106', created=1699053533, object='model', owned_by='system'),
 Model(id='text-embedding-3-small', created=1705948997, object='model', owned_by='system'),
 Model(id='tts-1-hd', created=1699046015, object='model', owned_by='system'),
 Model(id='gpt-4-0125-preview', created=1706037612, object='model', owned_by='system'),
 Model(id='davinci-002', created=1692634301, object='model', owned_by='system'),
 Model(id='babbage-002', created=1692634615, object='model', owned_by='system'),
 Model(id='gpt-4-turbo-preview', created=1706037777, object='model', owned_by='system'),
 Model(id='text-embedding-3-large', created=1705953180, object=

In [275]:
chat_llm = ChatOpenAI(
    model='gpt-3.5-turbo-1106',
    temperature=0
)

In [276]:
chat_llm.invoke('hi')

AIMessage(content='Hello! How can I assist you today?')

- prompt

In [277]:
prompt = """
You are a top financial lawyer with rich experience who has worked in the top law firm Magic Circle for 60 years. Next, you will answer the following QUESTION according to the relevant INFORMATION of a company, where you analyze a company's governance structure to determine the likelihood of a 'Directors102b7' waiver, which exculpates directors from monetary liability for breach of fiduciary duty of care. Note: this data is biased and if you're not sure about the prediction, lean on predicting T.

Consider the following governance features and their implications on director liability protections:
Given a company's governance structure, we want to predict the presence of a 'Directors102b7' waiver. If
the company has 'IndemnityDirector' policies, this suggests a protective
approach towards leadership liability, making a 'Directors102b7' waiver more likely. Additionally,
'InsuranceDirector' policies that require D&O insurance for directors further indicate a tendency to shield
 directors from financial risks. However, we must consider the broader governance context. For instance, a
 company with 'CumulativeVoting' rights may prioritize shareholder empowerment over director protections,
 potentially making 'Directors102b7' waivers less likely, even if other protective measures are in place.

QUESTION:
Based on the provided company's governance structure, does the charter exculpate directors from monetary liability for breach of fiduciary duty of care (a "102b7" waiver)?

INFORMATION:
{{content}}

Your analysis should lead to a conclusion that is logically derived from the governance features mentioned. Consider both the presence and absence of these features in your reasoning.

Your output should resemble a VALID JSON Object as illustrated below:
{
    "type": "Y", "N",
    "reason": "Provide a concise explanation based on the governance features analyzed."
}

- "Y" means the charter likely exculpates directors from monetary liability for breach of fiduciary duty of care.
- "N" means it is unlikely that the charter exculpates directors from such liability.

Let's think step by step to reach a reasoned conclusion.


FINAL ANSWER:
""".strip()

In [278]:
chat_llm = ChatOpenAI(
    model='gpt-3.5-turbo-1106',
    temperature=0.3
)

chat = (
    PromptTemplate.from_template(prompt, template_format='jinja2')
    |chat_llm
)

In [279]:
test = {**data[31]}
print(test)

test.pop('Directors102b7')
resp = chat.invoke({"content": json.dumps(test)})

print(resp.content)

{'CIK': '91576', 'Charter_ID': '91576A20190523', 'Date_Coded': '1998-03-04', 'Date_Filing': '1994-02-25', 'CompanyName': 'KEYCORP  NEW', 'State': 'OH', 'CumulativeVoting': 'Prohibit', 'Directors102b7': 'N', 'Officers102b7': 'N', 'IndemnityDirector': 'N', 'IndemnityOfficer': 'N'}
{
    "type": "N",
    "reason": "The company's governance structure prohibits cumulative voting and does not have indemnity director policies, indicating a lack of emphasis on director protections. Therefore, it is unlikely that the charter exculpates directors from monetary liability for breach of fiduciary duty of care."
}


In [280]:
data[31]['Directors102b7']

'N'

In [281]:


# from tqdm import tqdm

# labels = []
# predicts = []

# for d in tqdm(data_sub[:2]):
#     test = {**d}
#     label = test.pop('Directors102b7')

#     resp = chat.invoke({"content": json.dumps(test)})

#     try:
#         resp = json.loads(resp.content.replace('"','').replace("'",'"'))
#         print(f' label:{label} predict:{resp}')
#     except Exception as e:
#         print(e)
    
#     labels.append(label)
#     predicts.append(resp)


In [282]:
from tqdm import tqdm
import math
import time

labels = []
predicts = []

gn = 10  # The batch size
groups = math.ceil(len(data_sub) / gn)

for n in tqdm(range(groups)):
    # Select the subset of data for the current batch
    ds = data_sub[gn * n : gn * (n + 1)]
    # Extract the labels for "Directors102b7"
    labels_batch = [d["Directors102b7"] for d in ds]
    # Remove the "Directors102b7" from the data that will be sent for prediction
    for d in ds:
        d.pop("Directors102b7", None)  # Use pop to remove the key, if it doesn't exist it does nothing

    # Prepare the data for the batch request, keeping the relevant columns for prediction
    test = [{k: d[k] for k in ["CumulativeVoting", "IndemnityDirector", "IndemnityOfficer"] if k in d} for d in ds]

    resp = chat.batch([{"content": json.dumps(t)} for t in test])

    # Store the true labels and predictions
    labels.extend(labels_batch)
    predicts.extend(resp)

    # Sleep to respect potential API rate limits
    time.sleep(5)


100%|██████████| 25/25 [03:00<00:00,  7.23s/it]


In [283]:
json.loads(predicts[0].content)

{'type': 'N',
 'reason': "The company's governance structure does not include 'IndemnityDirector' policies or 'InsuranceDirector' policies, which would suggest a protective approach towards director liability. Additionally, the absence of 'CumulativeVoting' rights indicates a potential prioritization of shareholder empowerment over director protections, making it unlikely for a 'Directors102b7' waiver to be present in the charter."}

In [284]:
# predicts 

predicts_r = []
for pi in range(len(predicts)):
    p = predicts[pi].content

    try:
        p = json.loads(p)
        print(f" label:{labels[pi]} predict:{p}")
    except Exception as e:
        p = {"error": p}
        print(e)

    predicts_r.append(p)

 label:Y predict:{'type': 'N', 'reason': "The company's governance structure does not include 'IndemnityDirector' policies or 'InsuranceDirector' policies, which would suggest a protective approach towards director liability. Additionally, the absence of 'CumulativeVoting' rights indicates a potential prioritization of shareholder empowerment over director protections, making it unlikely for a 'Directors102b7' waiver to be present in the charter."}
 label:Y predict:{'type': 'Y', 'reason': "The company's governance structure includes 'IndemnityDirector' policies, which suggest a protective approach towards leadership liability, making a 'Directors102b7' waiver more likely. Additionally, the presence of 'IndemnityOfficer' policies further indicates a tendency to shield directors from financial risks. Although 'CumulativeVoting' rights may prioritize shareholder empowerment, the overall protective measures in place make it likely that the charter exculpates directors from monetary liabili

In [285]:
# labels[:10], predicts_r

In [286]:
from sklearn.metrics import accuracy_score, confusion_matrix

label_map = {
    'Y':1,
    'N':0
}

labels_logit = [label_map.get(i, 2) for i in labels]
predicts_logit = [label_map.get(i['type'], 2) for i in predicts_r]

accuracy_score(labels_logit, predicts_logit)

0.6585365853658537

In [287]:
confusion_matrix(labels_logit, predicts_logit)

array([[ 28,  13],
       [ 71, 134]], dtype=int64)

##  conclusion

   1. In the classification task focusing on the "Directors102b7" column with an additional "filtering" of related governance features such as "CumulativeVoting", "IndemnityDirector", and "IndemnityOfficer", and also a data cleaning for NAN values the model achieved an accuracy of approximately 65%. 
      The observation of the features that are related to Directors102b7 is also done with the help of a simpler CoT via ChatGPT.
   > confusion matrix
      [[  28,  13],
       [  71, 134]]
         

      


   
   2. Before, when we include "Officers102b7", which is a feature that is not so related to our result, the model achieved an accuracy of approximately 49.13%. This targeted approach makes the confusion matrix include substantial misclassifications. 
   
   > confusion matrix
      [[  0,  50,  50],
       [  0, 254, 161],
       [  0,  31,  28]]

   3. On the other hand, when the model was applied to a more extensive set of variables encompassing the entire dataset "without filtering", the accuracy significantly dropped to around 18.12%. 
   
   
   > the other matrix
      [[ 81,   0,   0],
       [411,  21,   0],
       [ 59,   0,   2]]
     
   The contrast and improvement of the three ways shows the critical importance of pre-processing data and also wisely choosing the related feature in data manipulating.
